```
Error running mint automation
+ mkdir -p 10355-0a755d3/gopath/src/github.com/minio
+ git -C 10355-0a755d3/gopath/src/github.com/minio clone --quiet https://github.com/minio/minio.git
+ git -C 10355-0a755d3/gopath/src/github.com/minio/minio remote add minio https://github.com/minio/minio.git
+ git -C 10355-0a755d3/gopath/src/github.com/minio/minio fetch --quiet minio pull/10355/head:pr10355
+ git -C 10355-0a755d3/gopath/src/github.com/minio/minio checkout --quiet pr10355
+ GO111MODULE=on
+ GOPATH=/home/mint-auto/mint-auto/10355-0a755d3/gopath
+ make -C 10355-0a755d3/gopath/src/github.com/minio/minio --quiet
go: downloading github.com/minio/cli v1.22.0
go: downloading github.com/json-iterator/go v1.1.10
go: downloading github.com/Azure/azure-storage-blob-go v0.8.0
go: downloading github.com/minio/sha256-simd v0.1.1
go: downloading github.com/colinmarc/hdfs/v2 v2.1.1
go: downloading gopkg.in/jcmturner/gokrb5.v7 v7.3.0
go: downloading github.com/minio/minio-go/v7 v7.0.5-0.20200811211821-14ed05478889
go: downloading github.com/klauspost/readahead v1.3.1
go: downloading github.com/prometheus/client_golang v1.0.0
go: downloading github.com/dchest/siphash v1.2.1
go: downloading github.com/minio/selfupdate v0.3.1
go: downloading github.com/google/uuid v1.1.1
go: downloading github.com/cespare/xxhash/v2 v2.1.1
go: downloading github.com/klauspost/reedsolomon v1.9.9
go: downloading golang.org/x/net v0.0.0-20200707034311-ab3426394381
go: downloading github.com/nsqio/go-nsq v1.0.7
go: downloading github.com/shirou/gopsutil v2.20.3-0.20200314133625-53cec6b37e6a+incompatible
go: downloading google.golang.org/api v0.5.0
go: downloading github.com/djherbis/atime v1.0.0
go: downloading github.com/xdg/scram v0.0.0-20180814205039-7eeb5667e42c
go: downloading golang.org/x/crypto v0.0.0-20200709230013-948cd5f35899
go: downloading github.com/go-sql-driver/mysql v1.5.0
go: downloading github.com/rjeczalik/notify v0.9.2
go: downloading go.etcd.io/etcd/v3 v3.3.0-rc.0.0.20200707003333-58bb8ae09f8e
go: downloading github.com/Azure/azure-pipeline-go v0.2.1
go: downloading github.com/pkg/errors v0.9.1
go: downloading github.com/modern-go/reflect2 v1.0.1
go: downloading gopkg.in/ldap.v3 v3.0.3
go: downloading github.com/prometheus/client_model v0.0.0-20190812154241-14fe0d1b01d4
go: downloading github.com/klauspost/compress v1.10.3
go: downloading github.com/gomodule/redigo v2.0.0+incompatible
go: downloading github.com/cheggaaa/pb v1.0.28
go: downloading github.com/willf/bloom v2.0.3+incompatible
go: downloading github.com/klauspost/cpuid v1.3.1
go: downloading golang.org/x/sys v0.0.0-20200625212154-ddb9806d33ae
go: downloading github.com/dgrijalva/jwt-go v3.2.0+incompatible
go: downloading github.com/secure-io/sio-go v0.3.0
go: downloading github.com/golang/protobuf v1.4.0
go: downloading github.com/prometheus/common v0.4.1
go: downloading github.com/montanaflynn/stats v0.5.0
go: downloading github.com/streadway/amqp v0.0.0-20190404075320-75d898a42a94
go: downloading google.golang.org/protobuf v1.22.0
go: downloading github.com/minio/highwayhash v1.0.0
go: downloading cloud.google.com/go v0.39.0
go: downloading github.com/gogo/protobuf v1.3.1
go: downloading github.com/gorilla/rpc v1.2.0
go: downloading google.golang.org/grpc v1.26.0
go: downloading gopkg.in/asn1-ber.v1 v1.0.0-20181015200546-f715ec2f112d
go: downloading github.com/rs/cors v1.7.0
go: downloading github.com/spaolacci/murmur3 v1.1.0
go: downloading github.com/nats-io/stan.go v0.7.0
go: downloading golang.org/x/text v0.3.3
go: downloading github.com/beorn7/perks v1.0.0
go: downloading github.com/mitchellh/go-homedir v1.1.0
go: downloading github.com/dustin/go-humanize v1.0.0
go: downloading github.com/lib/pq v1.7.0
go: downloading github.com/elazarl/go-bindata-assetfs v1.0.0
go: downloading github.com/Shopify/sarama v1.24.1
go: downloading github.com/nats-io/nuid v1.0.1
go: downloading github.com/ncw/directio v1.0.5
go: downloading github.com/prometheus/procfs v0.0.8
go: downloading github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd
go: downloading github.com/minio/sio v0.2.0
go: downloading github.com/xdg/stringprep v1.0.0
go: downloading github.com/golang/snappy v0.0.1
go: downloading github.com/valyala/tcplisten v0.0.0-20161114210144-ceec8f93295a
go: downloading gopkg.in/ini.v1 v1.57.0
go: downloading github.com/willf/bitset v1.1.11
go: downloading go.uber.org/zap v1.14.1
go: downloading github.com/eapache/go-xerial-snappy v0.0.0-20180814174437-776d5712da21
go: downloading github.com/miekg/dns v1.1.8
go: downloading github.com/minio/md5-simd v1.1.0
go: downloading github.com/mattn/go-ieproxy v0.0.1
go: downloading github.com/googleapis/gax-go/v2 v2.0.4
go: downloading google.golang.org/genproto v0.0.0-20190819201941-24fa4b261c55
go: downloading github.com/rcrowley/go-metrics v0.0.0-20181016184325-3113b8401b8a
go: downloading github.com/gorilla/mux v1.8.0
go: downloading github.com/mattn/go-colorable v0.1.4
go: downloading github.com/matttproud/golang_protobuf_extensions v1.0.1
go: downloading github.com/tinylib/msgp v1.1.2
go: downloading github.com/pierrec/lz4 v2.4.0+incompatible
go: downloading github.com/fatih/color v1.7.0
go: downloading github.com/eclipse/paho.mqtt.golang v1.2.0
go: downloading golang.org/x/oauth2 v0.0.0-20190226205417-e64efc72b421
go: downloading github.com/mattn/go-runewidth v0.0.2
go: downloading github.com/hashicorp/go-uuid v1.0.1
go: downloading github.com/hashicorp/vault/api v1.0.4
go: downloading go.uber.org/multierr v1.5.0
go: downloading github.com/coreos/go-semver v0.2.0
go: downloading github.com/gorilla/handlers v1.4.2
go: downloading github.com/hashicorp/go-cleanhttp v0.5.1
go: downloading github.com/mattn/go-isatty v0.0.8
go: downloading github.com/hashicorp/go-retryablehttp v0.5.4
go: downloading github.com/davecgh/go-spew v1.1.1
go: downloading gopkg.in/olivere/elastic.v5 v5.0.86
go: downloading github.com/hashicorp/hcl v1.0.0
go: downloading golang.org/x/time v0.0.0-20190308202827-9d24e82272b4
go: downloading github.com/eapache/queue v1.1.0
go: downloading github.com/hashicorp/errwrap v1.0.0
go: downloading github.com/hashicorp/vault/sdk v0.1.13
go: downloading github.com/hashicorp/go-rootcerts v1.0.1
go: downloading go.opencensus.io v0.22.3
go: downloading git.apache.org/thrift.git v0.13.0
go: downloading github.com/klauspost/pgzip v1.2.1
go: downloading gopkg.in/jcmturner/rpc.v1 v1.1.0
go: downloading github.com/eapache/go-resiliency v1.1.0
go: downloading github.com/rs/xid v1.2.1
go: downloading github.com/minio/simdjson-go v0.1.5
go: downloading github.com/philhofer/fwd v1.0.0
go: downloading github.com/beevik/ntp v0.2.0
go: downloading github.com/tidwall/gjson v1.3.5
go: downloading github.com/jcmturner/gofork v0.0.0-20190328161633-dc7c13fece03
go: downloading gopkg.in/jcmturner/goidentity.v3 v3.0.0
go: downloading gopkg.in/square/go-jose.v2 v2.3.1
go: downloading gopkg.in/jcmturner/aescts.v1 v1.0.1
go: downloading github.com/hashicorp/go-multierror v1.0.0
go: downloading github.com/mailru/easyjson v0.7.1
go: downloading github.com/coredns/coredns v1.4.0
go: downloading github.com/hashicorp/go-sockaddr v1.0.2
go: downloading github.com/tidwall/sjson v1.0.4
go: downloading github.com/bcicen/jstream v0.0.0-20190220045926-16c1f8af81c2
go: downloading gopkg.in/yaml.v2 v2.2.8
go: downloading github.com/alecthomas/participle v0.2.1
go: downloading github.com/nats-io/nats.go v1.10.0
go: downloading gopkg.in/jcmturner/dnsutils.v1 v1.0.1
go: downloading github.com/coreos/go-systemd/v22 v22.0.0
go: downloading github.com/golang/groupcache v0.0.0-20190702054246-869f871628b6
go: downloading github.com/mitchellh/mapstructure v1.1.2
go: downloading github.com/tidwall/pretty v1.0.0
go: downloading github.com/fatih/structs v1.1.0
go: downloading go.uber.org/atomic v1.6.0
go: downloading github.com/tidwall/match v1.0.1
go: downloading github.com/ryanuber/go-glob v1.0.0
go: downloading github.com/nats-io/jwt v0.3.2
go: downloading github.com/nats-io/nkeys v0.1.4
import cycle not allowed
package github.com/minio/minio
	imports github.com/minio/minio/cmd
	imports github.com/minio/minio/cmd/gateway/azure
	imports github.com/minio/minio/cmd
make: *** [build] Error 1

```

--- a/cmd/bucket-metadata-sys.go
+++ b/cmd/bucket-metadata-sys.go
@@ -81,7 +81,7 @@ func (sys *BucketMetadataSys) Update(bucket string, configFile string, configDat
 		// This code is needed only for gateway implementations.
 		switch configFile {
 		case bucketSSEConfig:
-			if globalGatewayName == "nas" {
+			if globalGatewayName == NASBackendGateway {
 				meta, err := loadBucketMetadata(GlobalContext, objAPI, bucket)
 				if err != nil {
 					return err
@@ -90,7 +90,7 @@ func (sys *BucketMetadataSys) Update(bucket string, configFile string, configDat
 				return meta.Save(GlobalContext, objAPI)
 			}
 		case bucketLifecycleConfig:
-			if globalGatewayName == "nas" {
+			if globalGatewayName == NASBackendGateway {
 				meta, err := loadBucketMetadata(GlobalContext, objAPI, bucket)
 				if err != nil {
 					return err
@@ -99,7 +99,7 @@ func (sys *BucketMetadataSys) Update(bucket string, configFile string, configDat
 				return meta.Save(GlobalContext, objAPI)
 			}
 		case bucketTaggingConfig:
-			if globalGatewayName == "nas" {
+			if globalGatewayName == NASBackendGateway {
 				meta, err := loadBucketMetadata(GlobalContext, objAPI, bucket)
 				if err != nil {
 					return err
@@ -108,7 +108,7 @@ func (sys *BucketMetadataSys) Update(bucket string, configFile string, configDat
 				return meta.Save(GlobalContext, objAPI)
 			}
 		case bucketNotificationConfig:
-			if globalGatewayName == "nas" {
+			if globalGatewayName == NASBackendGateway {
 				meta, err := loadBucketMetadata(GlobalContext, objAPI, bucket)
 				if err != nil {
 					return err
@@ -273,7 +273,7 @@ func (sys *BucketMetadataSys) GetLifecycleConfig(bucket string) (*lifecycle.Life
 // GetNotificationConfig returns configured notification config
 // The returned object may not be modified.
 func (sys *BucketMetadataSys) GetNotificationConfig(bucket string) (*event.Config, error) {
-	if globalIsGateway && globalGatewayName == "nas" {
+	if globalIsGateway && globalGatewayName == NASBackendGateway {
 		// Only needed in case of NAS gateway.
 		objAPI := newObjectLayerWithoutSafeModeFn()
 		if objAPI == nil {

--- a/cmd/gateway-interface.go
+++ b/cmd/gateway-interface.go
@@ -21,7 +21,14 @@ import (
 )
 
 // GatewayMinioSysTmp prefix is used in Azure/GCS gateway for save metadata sent by Initialize Multipart Upload API.
-const GatewayMinioSysTmp = "minio.sys.tmp/"
+const (
+	GatewayMinioSysTmp  = "minio.sys.tmp/"
+	AzureBackendGateway = "azure"
+	GCSBackendGateway   = "gcs"
+	HDFSBackendGateway  = "hdfs"
+	NASBackendGateway   = "nas"
+	S3BackendGateway    = "s3"
+)
 
 // Gateway represents a gateway backend.
 type Gateway interface {

--- a/cmd/gateway-main.go
+++ b/cmd/gateway-main.go
@@ -262,8 +262,8 @@ func StartGateway(ctx *cli.Context, gw Gateway) {
 	}
 
 	// Currently only NAS and S3 gateway support encryption headers.
-	encryptionEnabled := gatewayName == "s3" || gatewayName == "nas"
-	allowSSEKMS := gatewayName == "s3" // Only S3 can support SSE-KMS (as pass-through)
+	encryptionEnabled := gatewayName == S3BackendGateway || gatewayName == NASBackendGateway
+	allowSSEKMS := gatewayName == S3BackendGateway // Only S3 can support SSE-KMS (as pass-through)
 
 	// Add API router.
 	registerAPIRouter(router, encryptionEnabled, allowSSEKMS)
@@ -310,7 +310,7 @@ func StartGateway(ctx *cli.Context, gw Gateway) {
 	// Calls all New() for all sub-systems.
 	newAllSubsystems()
 
-	if gatewayName == "nas" {
+	if gatewayName == NASBackendGateway {
 		buckets, err := newObject.ListBuckets(GlobalContext)
 		if err != nil {
 			logger.Fatal(err, "Unable to list buckets")

--- a/cmd/gateway/azure/gateway-azure.go
+++ b/cmd/gateway/azure/gateway-azure.go
@@ -73,7 +73,6 @@ const (
 	azureBlockSize             = 100 * humanize.MiByte
 	azureS3MinPartSize         = 5 * humanize.MiByte
 	metadataObjectNameTemplate = minio.GatewayMinioSysTmp + "multipart/v1/%s.%x/azure.json"
-	azureBackend               = "azure"
 	azureMarkerPrefix          = "{minio}"
 	metadataPartNamePrefix     = minio.GatewayMinioSysTmp + "multipart/v1/%s.%x"
 	maxPartsCount              = 10000
@@ -113,7 +112,7 @@ EXAMPLES:
 `
 
 	minio.RegisterGatewayCommand(cli.Command{
-		Name:               azureBackend,
+		Name:               minio.AzureBackendGateway,
 		Usage:              "Microsoft Azure Blob Storage",
 		Action:             azureGatewayMain,
 		CustomHelpTemplate: azureGatewayTemplate,
@@ -170,7 +169,7 @@ type Azure struct {
 
 // Name implements Gateway interface.
 func (g *Azure) Name() string {
-	return azureBackend
+	return minio.AzureBackendGateway
 }
 
 // NewGatewayLayer initializes azure blob storage client and returns AzureObjects.

--- a/cmd/gateway/gcs/gateway-gcs.go
+++ b/cmd/gateway/gcs/gateway-gcs.go
@@ -91,8 +91,6 @@ const (
 
 	// Project ID key in credentials.json
 	gcsProjectIDKey = "project_id"
-
-	gcsBackend = "gcs"
 )
 
 func init() {
@@ -132,7 +130,7 @@ EXAMPLES:
 `
 
 	minio.RegisterGatewayCommand(cli.Command{
-		Name:               gcsBackend,
+		Name:               minio.GCSBackendGateway,
 		Usage:              "Google Cloud Storage",
 		Action:             gcsGatewayMain,
 		CustomHelpTemplate: gcsGatewayTemplate,
@@ -145,13 +143,13 @@ func gcsGatewayMain(ctx *cli.Context) {
 	projectID := ctx.Args().First()
 	if projectID == "" && os.Getenv("GOOGLE_APPLICATION_CREDENTIALS") == "" {
 		logger.LogIf(minio.GlobalContext, errGCSProjectIDNotFound, logger.Application)
-		cli.ShowCommandHelpAndExit(ctx, "gcs", 1)
+		cli.ShowCommandHelpAndExit(ctx, minio.GCSBackendGateway, 1)
 	}
 	if projectID != "" && !isValidGCSProjectIDFormat(projectID) {
 		reqInfo := (&logger.ReqInfo{}).AppendTags("projectID", ctx.Args().First())
 		contxt := logger.SetReqInfo(minio.GlobalContext, reqInfo)
 		logger.LogIf(contxt, errGCSInvalidProjectID, logger.Application)
-		cli.ShowCommandHelpAndExit(ctx, "gcs", 1)
+		cli.ShowCommandHelpAndExit(ctx, minio.GCSBackendGateway, 1)
 	}
 
 	minio.StartGateway(ctx, &GCS{projectID})
@@ -164,7 +162,7 @@ type GCS struct {
 
 // Name returns the name of gcs ObjectLayer.
 func (g *GCS) Name() string {
-	return gcsBackend
+	return minio.GCSBackendGateway
 }
 
 // NewGatewayLayer returns gcs ObjectLayer.

--- a/cmd/gateway/hdfs/gateway-hdfs.go
+++ b/cmd/gateway/hdfs/gateway-hdfs.go
@@ -47,8 +47,6 @@ import (
 )
 
 const (
-	hdfsBackend = "hdfs"
-
 	hdfsSeparator = minio.SlashSeparator
 )
 
@@ -84,7 +82,7 @@ EXAMPLES:
 `
 
 	minio.RegisterGatewayCommand(cli.Command{
-		Name:               hdfsBackend,
+		Name:               minio.HDFSBackendGateway,
 		Usage:              "Hadoop Distributed File System (HDFS)",
 		Action:             hdfsGatewayMain,
 		CustomHelpTemplate: hdfsGatewayTemplate,
@@ -96,7 +94,7 @@ EXAMPLES:
 func hdfsGatewayMain(ctx *cli.Context) {
 	// Validate gateway arguments.
 	if ctx.Args().First() == "help" {
-		cli.ShowCommandHelpAndExit(ctx, hdfsBackend, 1)
+		cli.ShowCommandHelpAndExit(ctx, minio.HDFSBackendGateway, 1)
 	}
 
 	minio.StartGateway(ctx, &HDFS{args: ctx.Args()})
@@ -109,7 +107,7 @@ type HDFS struct {
 
 // Name implements Gateway interface.
 func (g *HDFS) Name() string {
-	return hdfsBackend
+	return minio.HDFSBackendGateway
 }
 
 func getKerberosClient() (*krb.Client, error) {

--- a/cmd/gateway/nas/gateway-nas.go
+++ b/cmd/gateway/nas/gateway-nas.go
@@ -24,10 +24,6 @@ import (
 	"github.com/minio/minio/pkg/auth"
 )
 
-const (
-	nasBackend = "nas"
-)
-
 func init() {
 	const nasGatewayTemplate = `NAME:
   {{.HelpName}} - {{.Usage}}
@@ -61,7 +57,7 @@ EXAMPLES:
 `
 
 	minio.RegisterGatewayCommand(cli.Command{
-		Name:               nasBackend,
+		Name:               minio.NASBackendGateway,
 		Usage:              "Network-attached storage (NAS)",
 		Action:             nasGatewayMain,
 		CustomHelpTemplate: nasGatewayTemplate,
@@ -73,7 +69,7 @@ EXAMPLES:
 func nasGatewayMain(ctx *cli.Context) {
 	// Validate gateway arguments.
 	if !ctx.Args().Present() || ctx.Args().First() == "help" {
-		cli.ShowCommandHelpAndExit(ctx, nasBackend, 1)
+		cli.ShowCommandHelpAndExit(ctx, minio.NASBackendGateway, 1)
 	}
 
 	minio.StartGateway(ctx, &NAS{ctx.Args().First()})
@@ -86,7 +82,7 @@ type NAS struct {
 
 // Name implements Gateway interface.
 func (g *NAS) Name() string {
-	return nasBackend
+	return minio.NASBackendGateway
 }
 
 // NewGatewayLayer returns nas gatewaylayer.

--- a/cmd/gateway/s3/gateway-s3.go
+++ b/cmd/gateway/s3/gateway-s3.go
@@ -40,10 +40,6 @@ import (
 	"github.com/minio/minio/pkg/bucket/policy"
 )
 
-const (
-	s3Backend = "s3"
-)
-
 func init() {
 	const s3GatewayTemplate = `NAME:
   {{.HelpName}} - {{.Usage}}
@@ -76,7 +72,7 @@ EXAMPLES:
 `
 
 	minio.RegisterGatewayCommand(cli.Command{
-		Name:               s3Backend,
+		Name:               minio.S3BackendGateway,
 		Usage:              "Amazon Simple Storage Service (S3)",
 		Action:             s3GatewayMain,
 		CustomHelpTemplate: s3GatewayTemplate,
@@ -109,7 +105,7 @@ type S3 struct {
 
 // Name implements Gateway interface.
 func (g *S3) Name() string {
-	return s3Backend
+	return minio.S3BackendGateway
 }
 
 const letterBytes = "abcdefghijklmnopqrstuvwxyz01234569"

--- a/cmd/metrics.go
+++ b/cmd/metrics.go
@@ -157,7 +157,7 @@ func healingMetricsPrometheus(ch chan<- prometheus.Metric) {
 // collects gateway specific metrics for MinIO instance in Prometheus specific format
 // and sends to given channel
 func gatewayMetricsPrometheus(ch chan<- prometheus.Metric) {
-	if !globalIsGateway || (globalGatewayName != "s3" && globalGatewayName != "azure" && globalGatewayName != "gcs") {
+	if !globalIsGateway || (globalGatewayName != S3BackendGateway && globalGatewayName != AzureBackendGateway && globalGatewayName != GCSBackendGateway) {
 		return
 	}
 

